### PR TITLE
test: Scalable Push Queries: RQTT tests

### PR DESCRIPTION
**Pull request by AlanConfluent**
### Description 
Adds a bunch of RQTT tests to Scalable push queries.  In order to do that, I had to add support for `"auto.offset.reset" : "latest"` since normally the input is produced first before the query -- this required producing the input after the query has been issued.  Also, to prevent a race, I required the framework wait for the response header so it's known that the query is running and ready to receive input.

These tests don't include tombstones yet since the code don't support them well.  That will be for a followup.

### Testing done 
These are all tests!

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/scalable-push-queries.json [hunk 1]
```diff
@@ -0,0 +1,773 @@
...
+        {"topic": "test_topic", "timestamp": 12345, "key": "11", "value": {"id": 100}},
+        {"topic": "test_topic", "timestamp": 12365, "key": "12", "value": {"id": 101}}
+      ],
+      "inputConditions": {
+        "waitForQueryHeader": true
+      },
+      "properties": {
+        "auto.offset.reset" : "latest"
+      },
+      "responses": [
```
Comment (agavra) [inline ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/scalable-push-queries.json:23]: is there any indication that it indeed issued a scalable push query and not just a regular old push query? we should test for that as well in case of regression

File: ksqldb-functional-tests/src/test/resources/test-runner/correct_foo/multi_join/input.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (agavra) [inline ksqldb-functional-tests/src/test/resources/test-runner/correct_foo/multi_join/input.json:1]: ?

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java [hunk 3]
```diff
@@ -114,6 +114,7 @@ public void write(final OutputStream out) {
   }
 
   private void write(final OutputStream output, final StreamedRow row) throws IOException {
+    System.out.println("Writing row: " + row);
```
Comment (agavra) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java:117]: did you mean to commit this?

**Review by agavra — COMMENTED** [review]
Did a quick pass to get you some early feedback

File: ksqldb-engine/src/main/java/io/confluent/ksql/physical/scalablepush/locator/AllHostsLocator.java [hunk 4]
```diff
@@ -51,6 +52,7 @@ public AllHostsLocator(final Supplier<List<PersistentQueryMetadata>> allPersiste
     }
 
     return currentQueries.stream()
+        .filter(persistentQueryMetadata -> persistentQueryMetadata.getState() == State.RUNNING)
```
[the remark this answers is not in the document]
Comment (AlanConfluent) [reply ksqldb-engine/src/main/java/io/confluent/ksql/physical/scalablepush/locator/AllHostsLocator.java:55]: This only is used at the moment at the start of the query to know which nodes we should contact.  The idea is that at the moment, if the query isn't running anywhere yet, we should get an error.  This works in RQTT at the moment so that we can retry many times until everything is up and running.  The alternative is that we can block and poll the state until it's running, but that seems less ideal.  We might want to also try to fail until it's running everywhere but that's beyond what this change does.

File: ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java [hunk 5]
```diff
@@ -295,10 +313,21 @@ private static StatementSplit splitStatements(final RestTestCase testCase) {
 
   private List<RqttResponse> sendQueryStatements(
       final RestTestCase testCase,
-      final List<String> statements
+      final List<String> statements,
+      final Optional<Runnable> afterHeader
   ) {
+    // We only produce inputs after the first query at the moment to simplify things
+    final boolean[] runAfterHeader = new boolean[1];
```
[the remark this answers is not in the document]
Comment (AlanConfluent) [reply ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java:320]: All of the ones I added have just one query.  It seems reasonable to me to throw an error in this case.

File: ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java [hunk 6]
```diff
@@ -319,6 +348,57 @@ private static StatementSplit splitStatements(final RestTestCase testCase) {
...
+      header.get();
+    } catch (Exception e) {
+      LOG.error("Error awaiting header", e);
+      throw new RuntimeException(e);
+    }
+
+    try {
+      afterHeader.run();
+    } catch (Exception e) {
+      LOG.error("Error doing after header", e);
```
[the remark this answers is not in the document]
Comment (AlanConfluent) [reply ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java:387]: Nope, this was an oversight.  I collapse these three try blocks into one since there's no good reason they should be separate.  I think I had done it to have a slightly better error message, but it should be obvious from the stacktrace.

File: ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java [hunk 7]
```diff
@@ -319,6 +348,57 @@ private static StatementSplit splitStatements(final RestTestCase testCase) {
...
+    }
+
+    try {
+      return Optional.of(future.get());
+    } catch (Exception e) {
+      LOG.error("Error awaiting rows", e);
+      throw new RuntimeException(e);
+    } finally {
+      subscriber.close();
+      publisher.close();
```
[the remark this answers is not in the document]
Comment (AlanConfluent) [reply ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java:397]: Collapsed to this one block, so it should always happen.

File: ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java [hunk 8]
```diff
@@ -319,6 +348,57 @@ private static StatementSplit splitStatements(final RestTestCase testCase) {
...
+    }
+//    return Optional.empty();
```
[the remark this answers is not in the document]
Comment (AlanConfluent) [reply ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java:399]: Done.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/alan/collect-list.json [hunk 9]
```diff
@@ -0,0 +1,241 @@
+{
+  "comments": [
+    "You can specify multiple statements per test case, i.e., to set up the various streams needed",
```
[the remark this answers is not in the document]
Comment (AlanConfluent) [reply ksqldb-functional-tests/src/test/resources/query-validation-tests/alan/collect-list.json:3]: You caught an early rough version of this PR where I accidentally included some unrelated files.  I removed them.

File: ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestQueryTranslationTest.java [hunk 10]
```diff
@@ -73,7 +73,7 @@
 @RunWith(Parameterized.class)
 public class RestQueryTranslationTest {
 
-  private static final Path TEST_DIR = Paths.get("rest-query-validation-tests");
+  private static final Path TEST_DIR = Paths.get("rest-query-validation-tests/alan");
```
[the remark this answers is not in the document]
Comment (AlanConfluent) [reply ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestQueryTranslationTest.java]: Nope, this was for testing, didn't mean to include it.

File: ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/scalable-push-queries.json (hunk 1 above)
Comment (AlanConfluent) [reply ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/scalable-push-queries.json:23]: I was debating how to do that.  The queryid is prefixed by SCALABLE_PUSH_.  That might be a decent way to check this.  Alternatively, there could be a force SPQ flag that would error the query if it wasn't an SQP, but that seems like not that right approach since they're meant to happen transparently.

File: ksqldb-functional-tests/src/test/resources/test-runner/correct_foo/multi_join/input.json (hunk 2 above)
Comment (AlanConfluent) [reply ksqldb-functional-tests/src/test/resources/test-runner/correct_foo/multi_join/input.json]: Removed.  Was included by accident.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java (hunk 3 above)
Comment (AlanConfluent) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java:117]: Nope!

File: ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/scalable-push-queries.json (hunk 1 above)
Comment (agavra) [reply ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/scalable-push-queries.json:23]: for now I think using the `queryID` is OK, but we should think about a way to expose to the user whether what they issued was scalable or not (one option could be some kind of `EXPLAIN` for push queries) then our framework can always issue the explain before issuing the query and fail if it isn't scalable. 

Can you create tickets to track this (one for exposing what type of push query it was to the user and one for updating RQTTs to leverage that)?

File: ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java [hunk 11]
```diff
@@ -535,6 +615,45 @@ private void waitForWarmStateStore(
...
+        LOG.info("Server responded with an error code to a scalable push query. "
+            + "This could be because the persistent query hasn't started yet"
+            + System.lineSeparator() + errorMessage);
+        threadYield();
+      } else {
+        resp.getResponse().close();
+        return;
+      }
+    }
+    LOG.info("Timed out waiting for non error");
```
[the remark this answers is not in the document]
Comment (AlanConfluent) [reply ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java:650]: I had to add a new field with streams state to be able to do this, but I did.  Take a look at the update.  I think it's less brittle.

File: ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/scalable-push-queries.json (hunk 1 above)
Comment (AlanConfluent) [reply ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/scalable-push-queries.json:23]: I added the ability to check the queryId prefix and am able to verify it's a scalable push query.

File: ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/scalable-push-queries.json [hunk 12]
```diff
@@ -24,7 +24,7 @@
         {"admin": {"@type": "currentStatus"}},
         {"admin": {"@type": "currentStatus"}},
         {"query": [
-          {"header":{"schema":"`K` STRING"}},
+          {"header":{"schema":"`K` STRING", "queryId|starts_with": "SCALABLE_PUSH_QUERY"}},
```
Comment (agavra) [inline ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/scalable-push-queries.json:27]: clever including this in the header ;)

File: ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java [hunk 13]
```diff
@@ -535,6 +613,43 @@ private void waitForWarmStateStore(
...
+    final long threshold = System.currentTimeMillis() + MAX_STATIC_WARM_UP.toMillis();
+    while (System.currentTimeMillis() < threshold) {
+      final RestResponse<KsqlEntityList> resp;
+      try {
+        resp = restClient.makeKsqlRequest("SHOW QUERIES;");
+      } catch (Throwable t) {
+        t.printStackTrace();
+        throw t;
+      }
+      if (resp.isErroneous()) {
```
Comment (agavra) [inline ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/rest/RestTestExecutor.java:627]: nit: we can move this out of the while loop now, right? (and the error below would no longer be accurate)

File: ksqldb-rest-model/src/main/java/io/confluent/ksql/rest/entity/RawQueryStatusCount.java [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.Objects;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.KafkaStreams.State;
+
+/**
+ * Used to keep track of a the state of KafkaStreams application
+ * across multiple servers. Used in {@link RunningQuery}.
+ */
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class RawQueryStatusCount {
```
Comment (agavra) [inline ksqldb-rest-model/src/main/java/io/confluent/ksql/rest/entity/RawQueryStatusCount.java:19]: as discussed offline, let's move this to a different PR (and we can keep the while loop checking for non-errors for now) as this affects production code

**Review by agavra — APPROVED** [review]
LGTM, but let's separate the `RawQueryStatusCount` to a separate PR so we can reason about that separately

File: ksqldb-rest-model/src/main/java/io/confluent/ksql/rest/entity/RawQueryStatusCount.java (hunk 14 above)
Comment (AlanConfluent) [reply ksqldb-rest-model/src/main/java/io/confluent/ksql/rest/entity/RawQueryStatusCount.java]: Done
